extras/devices/LVFS/LVFS.py

```diff
@@ -0,0 +1,301 @@
+import subprocess
+import json
+import time
+from devices.base import AbstractDevice, logger
+
+SSH_OPTS = "-o StrictHostKeyChecking=no -o UserKnownHostsFile=/dev/null"
+
+
+class LVFSDevice(AbstractDevice):
+    fw_update_type = "LVFS"
+    vendor = ["HP", "Dell Inc.", "LENOVO"]
+    reboot_timeout = 600
+
+    def __init__(self, ipaddr, user, password):
+        super().__init__(ipaddr, user, password)
+
+    def run_cmd(self, cmd, raise_stderr=True, timeout=30):
+        """
+        Execute command on DUT via SSH
+
+        :returns: return code, stdout, stderr
+        :rtype: int, string, string
+        """
+        if self.password == "":
+            ssh_cmd = f'ssh -t {SSH_OPTS} {self.user}@{self.ipaddr} "{cmd}"'
+        else:
+            ssh_cmd = f'sshpass -p {self.password}  ssh -t {SSH_OPTS} {self.user}@{self.ipaddr} "{cmd}"'
+        logger.debug(f"Run command: {ssh_cmd}")
+        r = subprocess.run(
+            ssh_cmd,
+            shell=True,
+            capture_output=True,
+            timeout=timeout,
+        )
+        rc, stdout, stderr = (
+            r.returncode,
+            r.stdout.decode().strip(),
+            r.stderr.decode().strip(),
+        )
+        if raise_stderr and rc != 0:
+            err_msg = f"Failed to execute {cmd}:\n [{rc}] {stdout} {stderr}"
+            logger.debug(err_msg)
+            raise RuntimeError(err_msg)
+        return rc, stdout, stderr
+
+    def get_fw_info(self):
+        """
+        Get current firmware version of all updatable devices on DUT, and print
+        out devices with upgradable/downgradable versions
+        """
+        self._install_fwupd()
+        logger.info("collect firmware info")
+        self.run_cmd("sudo fwupdmgr refresh --force")
+        rc, stdout, stderr = self.run_cmd("sudo fwupdmgr get-devices --json")
+        logger.debug(f"$fwupdmgr get-devices = \n{stdout}")
```

Comment:
sorry, I should have been more specific about the f-strings. Logging is one place where you generally do NOT want to use f-strings, but use the built-in % syntax instead. It's weird, I know. Many times, it doesn't make a big difference, but it's a good habit to get into.

Reply:
I'm curious about this. Why is it better to use `%s` instead in logging?

Reply:
As I understand, the logging library is optimised to use the %s formatting style. I've found these links about it:
https://docs.python.org/3/howto/logging.html#optimization
https://docs.python.org/3/howto/logging-cookbook.html#formatting-styles

Reply:
Yes you're right. It is suggested to [use % syntax with logging](https://docs.python.org/3/howto/logging.html#optimization)`. Thanks for pointing out this!